---
title: "genomicSimulation Vignette"
author: "Kira Villiers"
date: "`r Sys.Date()`"
output:
  html_document:
      fig_caption: true
      number_sections: true
      toc: true
      theme: simplex
vignette: >
  \VignetteIndexEntry{gSvignette}
  \VignetteEngine{knitr::rmarkdown}
  \VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(genomicSimulation)
```

# Introduction
This package can be used to create 'digital twins' of breeding schemes. It offers scripting functions to simulate crossing and genomic selection on SNP markers. It was created with self-pollinated plant species in mind. 

The package itself is an interface to a set of C functions. Using C to perform the computation-intensive parts of simulation allows for quick runtimes - however, since this package is still mostly just a wrapper around those core C functions, it produces most of its output by writing to files rather than saving values to R.


# Details and Methodology

## Methodology: Calculating Genetic Estimated Breeding Values
The simulation loads the effect file into a set of vectors, each vector containing the effect values for each SNP for a particular allele. When the simulation is then instructed to calculate GEBVs for a genotype, it creates a vector containing how many copies of a particular allele the genotype contains at each SNP, then multiplies this vector with the effect vector for that allele. It repeats this process for each allele it has an effect vector for, then adds up the results to produce one final value.


## Methodology: Simulating Meiosis
To create a cross, the simulation generates gametes independently for the two parents. No distinction is made between male and female parent. 

For each chromosome number, the number of crossovers to occur in the simulated meiosis are drawn from a Poisson distribution with parameter = [length of the chromosome in Morgans], where length of the chromosome is pre-calculated as the distance between the first and last SNP that the `SimData` knows are in that chromosome. The positions of those crossovers along the chromosome are then drawn from a uniform distribution, which is why the use of a linkage map (so that distances correspond to likelihood of crossover) is recommended over use of a physical map. Finally, a random logical value (0 or 1) is drawn to choose which of the parent's chromosomes to start with. 

Note that this means that allele order in parents matters. <!--Encoded data loaders like `load_transposed_encoded_genes_to_simdata` randomise the order of alleles when the line is heterozygous at a SNP on loading.-->

The R package uses the R random generators, so will follow the R seed.

This procedure is displayed visually in the images below.

![Random draws for each chromosome](meiosisdists.png)

![Using those random values to generate a gamete](meiosis.png)


## Methodology: Counting Recombinations
Extra functionality is available for estimating where crossovers occured in producing certain genotypes. This is offered by the `find.crossovers` and `find.plot.crossovers` functions. The difference between the two is that after crossovers points are found, as well as saving the results matrix to a file `find.plot.crossovers` produces an R plot, as shown below. It assumes the child is mostly homozygous.

![An example plot produced by the crossover finder function with certainty=FALSE. The lines plotted were produced by crossing the two parents then selfing for 6 generations. This produces a large number of crossover events.](crossovers.png)

To identify crossovers, the code starts at the beginning of each chromosome for the current child line, and reads through to find SNPs where the parents differ (and so therefore the provenance of the allele the child has there can be determined).

The functions have two parameters that affect the way the crossover counting works: `window.size` and `certainty`.

If the `certainty` setting is `True`, then SNPs where the provenance of an allele cannot be determined are marked with a 0, representing unknown parentage. If the `certainty` setting is `False`, when the provenance of an allele is unknown, the parentage at that SNP is set to the same parent as the most recent SNP with known parentage.

The `window.size` parameter should be an odd positive integer. The window size is the number of SNPs that are observed at a time. Only if the alleles of the child at all SNPs in the window can come from one parent, and that same sequence cannot come from the other, is the parentage at that SNP considered known.



# Speed
On a consumer-model laptop running RStudio version 1.2.5019, the process of loading 50 founder lines and 5112 SNPs took on average 0.64 seconds.

A call to `cross.randomly` to generate 100,000 ($10^5$) offspring from those founders took on average 2 seconds to run. Since all crossing functions use the same methodology, this time should be fairly representative of the speed of any crossing function. 

Output (writing to files) adds significant delays. The same command, with save-as-you-go output of GEBVs and pedigrees requested, took approximately 20 seconds to run. With save-as-you-go output of GEBVs, pedigrees, and genotypes (the latter being a significantly larger file size), the function call took approximately 120 seconds. 

Generating the $10^5$ offspring, then saving GEBVs, pedigrees, and genotypes to files via separate `save.` functions, took approximately 130 seconds to complete. This is a few seconds longer than the save-as-you-go call that did the same work, possibly because of the time spent moving from R-level execution to C-level and back again.

The C-level functions make no attempt to parallelise execution.

# Package-Specific Details
## Understanding `sim.data`
This package's set of working data is stored a C struct called `SimData`. An external pointer to the struct is saved in a package-global R variable called `sim.data$p`. `sim.data$p` is not visible during regular use. 

To reset the simulation/clear it of all data, the function `clear.simdata()` is provided.

### Markers
Each `SimData` is given a list of SNPs and a map for positioning these in the genome on creation. These details cannot be changed after creation without clearing the `SimData`. The `SimData` object saves names of markers as-is, as long as the name is under 30 characters long. Behaviour is undefined if any marker name is longer than 30 characters.

Becauses the `SimData` stores its SNPs in genome order, and so any outputs produced will have the SNP markers ordered according to the genetic map loaded at its creation. 

### Effects
The simulation cannot run without having a genetic map for the SNPs it tracks. However, it can run with no effect values loaded, so in cases where there is no need to be using GEBVs, there is no need to provide an effect file.

Only one 'set' of effect values can be loaded at a time. That is, the simulation only has one way of calculating GEBVs at a time.

### Genotypes
Each genotype/line may or may not be allocated a name. If loading names from a file, behaviour is undefined if any line name is longer than 30 characters.

## Understanding Groups
The SimData allows for lines to be treated in groups/categories. Automatically, the lines loaded from separate files are allocated to separate groups. Each crossing function also places the lines it produces in a new group.

A line can only belong to one group, so functions that overwrite allocations (eg `make.group`) effectively pull lines out of their old group and put them in a new one. There's therefore no chance of attempting to delete lines via `delete.group` that have already been deleted.

Group numbers are natural numbers starting at 1. Group numbers are not guaranteed to be allocated sequentially, though functions usually choose the lowest currently-unused valid group number when creating their new groups.

Each separate function call that adds new genotypes to the internal `SimData` allocates them to a new, separate group. 

Every function except those that pull data into R (functions whose names begin with `see.`) return either one integer or a vector of integers. If the integer is 0, it means the function succeeded but it is not a function that acts on groups. Otherwise, the return value is the group number of the group(s) that were created.

## Understanding Identifiers
There are three categories of identifiers that may pop up during use of the simulation.

1. Group numbers, which were discussed in the section above. They are non-negative integers that define a set of lines. Any number of lines can share a group number, and a line's group number may be changed over the course of the simulation.
1. IDs. These are non-negative integers that are unique to a particular genotype for the lifetime of the `SimData`. IDs are used for tracking pedigree thanks to this uniqueness feature. However, genotypes are not guaranteed to have an ID: genotypes created during crossing with the setting `give.ids` set to FALSE will not be allocated IDs and will have 0 instead. 
1. Indexes. These represent the position/column number in the internal matrix at which the genotype is stored. They are therefore unique to the genotype but not continuous over the lifetime of the `SimData`, because when genotypes/columns are deleted the column numbers of the matrix change. Because, unlike for IDs, every genotype is guaranteed to have an index, functions like `make.group` and `cross` that act on particular genotypes generally take indexes as inputs. Since indexes can change, however, it is important to not place a `delete.group` call between getting and using indexes.

## Understanding Output Methods
There are three sets of functions to get output from the simulation. 

- The first involves calling a function to save data to a file (`save.` family functions). 
- The second involves saving the data from a cross as you go (`save.` settings in crossing function settings). This is useful in conjunction with the setting `retain=FALSE` when generating more genotypes than will fit in memory, because in save-as-you-go mode the simulation saves the data to a file every 1000 lines (and if `retain=FALSE` it can then delete them and free up memory). However, this method is more limited than full `save.` functions in that it always uses the default printing format for that data type and only prints the genotypes being generated by that crossing call. 
- The third (and most limited) is to use the `see.` family of functions to pull data into R. The family of `see.` functions will be expanded in future, but for now is fairly limited in scope.


# Workflow
The assumed workflow for this package involves the following steps:

1. Load progenitor lines and SNP data; save the group number of the progenitor lines. (If using GEBV functions: Load effect values for calculating GEBV for a desired trait). Suggested function: `load.data`
1. Perform crossing from the progenitor group. Keep a hold of the group number of the output. Example function: `cross.randomly`
1. Perform more crossing from the F1 generation (using the group number from the above step). Example.function: `cross.from.top.pc`
1. Repeat for further generations. Example function: `self.n.times`
1. Save the final genotypes (and/or effect values and/or pedigrees) to files. Example function: `save.genotypes`
1. (Optional: Estimate the number of crossovers to reach these final genotypes. Suggested function: `plot.crossovers`)
 Load output files into R for further analysis.

The example below shows a how a typical workflow would save group numbers and call crossing functions.
```{R}
# (Show the tiny example data set)
read.csv("a-matrix.txt", sep='\t', header=T)
read.csv("a-map.txt", sep='\t', header=T)
read.csv("a-yield.txt", sep='\t')

# load the small initial data set
g0 <- load.data("a-matrix.txt", "a-map.txt", "a-yield.txt")

# Do many random crosses from the data set. 
f1 <- cross.randomly(g0, n.crosses=20, give.names=TRUE, name.prefix="F1.")

# Let's see what crossovers we got. For this few markers we don't get much information.
save.pedigrees("a.txt", f1, type="P")
find.plot.crossovers("a.txt", "b.txt")

# Make 40 random crosses from the 25% with the top GEBV
f1.selected <- select.by.gebv(f1, percentage=25)

# Delete groups we are not currently using, to free up some memory.
delete.group(f1)

f2 <- cross.randomly(f1.selected, n.crosses=5)
delete.group(f1.selected)

# Complete 4 rounds of selfing.
f6 <- self.n.times(f2, 4, file.prefix="af6", save.genotype=TRUE)

# (Show current state of groups)
print(c(g0, f2, f6))
see.existing.groups()

# (Show these output files)
read.csv("af6-genome", sep='\t')

```

## Simple Example Breeding Scheme
![A demonstration of an implementation of a simplified breeding scheme. This process was run for two conditions, one with selection inside `get.top.by.phenotype` done across all individuals in the set, the other with selection done within each family of lines descended from one of the original 25 crosses.](example-scheme.png)

![The script snippet above was run for 10 replications of each condition, and the mean GEBV and variance in GEBV at the end of each cycle was tracked. The thin lines show individual replications and the thick ones the mean for that condition at that cycle. ](example-scheme-fig.png)

## Selection by Simulated Phenotype
```{R}
f <- cross.randomly(g0, n.crosses=5, offspring=3)

f.info <- see.group.gebvs(f)

# Create the masked values. In this case, a normally distributed term with sd corresponding
# to environmental variance is added to each GEBV
H2 <- 0.5  #estimate the heritability of the GEBV's trait at this step
Vg <- var(f.info$GEBV)
Ve <- Vg/H2 - Vg  # using the formula H2 = Vg/(Vg + Ve) to find the environmental variation
f.info$Pheno <- f.info$GEBV + rnorm(length(f.info$GEBV), mean=0, sd = sqrt(Ve))
  
# Sort the list by phenotypes
f.info <- f.info[order(f.info$Pheno, decreasing=TRUE),]
# Send the top 10 phenotypes to a new group
fselected <- make.group(f.info$i[1:10])

# (delete the non-selected genotypes)
delete.group(f)

# Then generate further generations from the selected genotypes
f2 <- cross.randomly(fselected, n.crosses=100)
# ...
delete.group(f2)
```


# Indexes
## Function Index
Use the R documentation (eg. type `?load.data` in the RStudio console) to get descriptions of these functions.

Loading data:

- `group <- load.data(allele.file, map.file, effect.file = NULL)`
- `group <- load.more.genotypes(allele.file)`
- `load.different.effects(effect.file)`

Simulating crossing:

- `group <- cross(parent1,parent2,[settings])`
- `group <- cross.randomly(group,n.crosses,[settings])`
- `group <- cross.combinations(cross.file,[settings])`
- `group <- cross.dc.combinations(cross.file,[settings])`
- `group <- cross.all.pairs(group,[settings])`
- `group <- self.n.times(group,n,[settings])`
- `group <- make.doubled.haploids(group,[settings])`

Performing selection:

- `selected.group <- select.by.gebv(from.group, low.score.best = FALSE, percentage OR number)`
- For more customised selection, use the process
  - `info <- see.group.gebvs(group)`
  - `selected.info <- `[some R manipulation on `info` that performs the custom selection]
  - `selected.group <- make.group(selected.info$I)`

Modify groupings:

- `delete.group(group)`
- `group <- combine.groups(groups)`
- `group <- make.group(indexes)`
- `[vector of groups] <- break.group.into.individuals(group)`
- `[vector of groups] <- break.group.into.families(group)`

Save data (saves results to files):

- `save.genome.model(filename)`
- `save.genotypes(filename, group = NULL, type = "R")`
- `save.allele.counts(filename, group = NULL, allele)`
- `save.pedigrees(filename, group = NULL, type = "R")`
- `save.GEBVs(filename, group = NULL)`
- `save.local.GEBVs(filename, block.file, group = NULL)`

View data (pulls data into R dataframes):

- `[dataframe containing current existing groups and their sizes] <- see.existing.groups()`
- `[vector containing the chosen data] <- see.group.data(group, data.type)`
- `[index and GEBV for each group member] <- see.group.gebvs(group)`

Identify recombination events:

- `find.crossovers(parentage.file, out.file, window.size = 1, certainty = TRUE)`
- `find.plot.crossovers(parentage.file, out.file, window.size = 1, certainty = TRUE)`

Data Storage:

- `clear.simdata()`

## Crossing Function Settings
All crossing functions take in a common set of settings. These are briefly described below.

- `offspring` : A setting representing the number of times to do each cross.
- `retain` : A boolean, repesenting whether to save the generated genotypes to memory or discard them. You may wish to discard them but save to file if you are generating too many crosses to save into memory.
- `give.names` : A boolean representing whether or not to produce names for the new genotypes generated. The names produced would have format [name.prefix][id]
- `name.prefix` : A string. If give.names is TRUE, the id is concatenated to this to produce the name of each new genotype.
- `track.pedigree` :A boolean representing whether or not to save the ids of the parents of each new genotype to the new genotype's pedigree. If this is false, the new genotype's pedigree is unknown.
- `give.ids` : A boolean representing whether or not to allocate each new genotype an id. If this is FALSE, the new genotype is 'invisible' to pedigree trackers and even if the pedigree of its offspring is supposedly tracked, the pedigree trackers will not be able to identify the progenitors of its offspring. Furthermore, if it is false and names are generated using give.names, all names generated in the same group will be the same. Probably you'd only have this FALSE if you were discarding the results or worried about id overflow.
- `file.prefix` : A string representing the prefix of files produced if save.pedigree=TRUE, save.gebv=TRUE, or save.genotype=TRUE.
- `save.pedigree` : A boolean. If TRUE, saves the pedigree in recursive format of each generated genotype to the text file with filename "[file.prefix]-pedigree".
- `save.gebv` : A boolean. If TRUE, saves the GEBVs of each generated genotype to the tab-separated text file with filename "[file.prefix]-eff".
- `save.genotype` : A boolean. If TRUE, saves the SNP/line matrix in regular format (generated genotypes as rows, SNPs as columns) to the tab-separated text file with filename "[file.prefix]-genome".


## Input File Formats
### Founder genotypes and SNPs
The basic file format expected to the `allele.file` parameter of `load.data` is:
```
name	G01	G02	G03	G04	G05	G06
m1	TT	TT	TT	TA	TT	AT
m3	TT	TT	TA	TA	TT	TT
m2	AA	AA	AA	AA	TT	AA
```
where G01, G02, ..., are the parent lines and m1, m2 and m3 are the SNPs. Cells are tab-separated. The value in the first cell ("name" in this example) is ignored.

To load more genotypes after original creation of the `SimData`, use the function `group <- load.more.genotypes(allele.file)`


### Linkage maps

The function's parameter `map.file` requires a file format like the following:
```
marker chr pos
m3 3 15
m2 1 8.3
m1 1 5.2
```
The first line's values are not checked. After that, it requires that all rows have three tab-separated values. The first should be the marker name, the second an integer representing the chromosome number, and the third a decimal representing the position of the SNP along the chromosome in centiMorgans (cM). 

This function loads only markers that already appear in the `SimData`'s SNP list, which is created from the SNPs in the genotype matrix file (`allele.file`). It will print out how many markers it successfully loaded.


### Effect files
Loading an effect file is optional for running the simulation. Its format should be:
```
m1 A -0.8
m2 A -0.1
m3 A 0.1
m1 T 0.9
m3 T -0.1
m2 T -0.5
```

The first column should be a SNP marker name, the second should be the allele this effect value corresponds to (as a single character), and then a decimal representing the effect value of that allele for that SNP.

### Combination crossing functions
Two crossing functions (`cross.combinations` and `cross.dc.combinations`) take instructions from files.

- `cross.combinations` : Take a file containing tab-separated pairs of line names and perform those crosses. An example file looks like:
```
G01	G02
G01	G03
G05	G06
G01	G04
```
- `cross.dc.combinations` : Supposes that only the names of the precursor lines are known but specific crosses between F1 individuals are wanted. Assumes that those F1 individuals have been produced with id allocation/pedigree tracking turned on. The function therefore reads in a tab-separated file that looks like the below to perform the cross between the offspring of G01 and G02, and the offspring of G03 and G05, and so forth.
```
G01	G02	G03	G05
G01	G03	G05	G06
G05	G06	G01	G04
```


## Output File Formats
### Saving genotypes
Use the function `save.genotypes` or the `save.genotype = TRUE` setting of crossing functions to save the alleles of generated genotypes across the list of SNPs.

It will save genotypes in the following format:
```
	m1	m2	m3
F107	TT	AA	TT
F108	TT	AA	TT
F109	TA	AA	TT
```

Note that this is the transpose of the usual expected input file format. Due to the linked list method used to store genotypes, it will be slower/take more passes over the data to print the transposed equivalent. However, this alternate format (shown below) is available via `save.genotypes` with `type = "T"`.
```
   F107   F108   F109
m1 TT     TT     TA
m2 AA     AA     AA
m3 TT     TT     TT
```

### Saving pedigrees
Pedigrees can be saved in one of two ways:

- `save.pedigrees`, with `type = "P"`, which saves in the following format, even if grandparents are known:
```
F2010   F107   F107
F2011   F108   F109
```
This is a tab-separated text file. The first column is the child name, and the next two columns are the names of its immediate parents (or single parent duplicated if it was produced by selfing or doubling haploids). If names are not available, IDs are used instead.

- `save.pedigrees`, with `type = "R"`, or crossing with setting `save.pedigree = TRUE`, which save in the following format:
```
10	F2010=(F107(G01,G02))
11	F2011=(F108(G01,G03),F109(G02,G03))
```
The function is recursive and will continue to track back to find parents until no pedigree is tracked. Note that F2010 is produced by selfing or doubling a haploid of F107, while F2011 is a cross.

### Saving GEBVs
Saving GEBVs, either via the `save.GEBVs` function or the `save.gebv = TRUE` setting of crossing functions, produces an output with the following format:
```
1	G01	1.400000
2	G02	1.400000
3	G03	1.600000
```
This is a tab-separated text file. The first column is the ID, then the name of each line, then the GEBV of that line.

### Saving block effect values/local GEBVs
The effect values for each parental half of a genotype's alleles and summed up for markers in a particular block (defined in the `block.file` parameter) can be saved to a file via the `save.local.GEBVs` function.

A block file with format the following format should be the second parameter:
```
Chrom	Pos	Name	Class	Markers
1	2.3	b000000	b	m1;m2;m3;
2 1.5 b000001 b m4;
...
```

The chromosome number, position and class columns are ignored. Any number of markers can be semicolon-separated in the last column. 

The output file has format:
```
    G1  G2  G3  G4 ...
b000000_1   0.25    0.25    -0.45   -0.45 ...
b000000_2   0.25    0.20    0.05    0.20 ...
b000001_1 ...
b000001_2 ...
```
This is a tab-separated text file.

### Save a summary of the `SimData`
The function `save.genome.model` is available to print out the map and effect values that the current `SimData` is using. The format in which it prints them is:
```
name	chr	pos	A	T
m1	1	5.200000	-0.800000	0.900000
m2	1	8.300000	-0.100000	-0.500000
m3	3	15.000000	0.100000	-0.100000
```
This is a tab-separated text file. 

### Recombination counters
The matrix saved to a file will have a format like the following:
```
	m1	m2	m3
F07	0	1	0
F08	0	2	0
F09	0	2	0
F10	0	1	0
```
A 0 refers to an unknown parentage. Non-zero values are the ID of the parent from which the allele was inherited. 


## R Output Formats

### `see.existing.groups`
This function returns a dataframe with two columns:

- `Group`, containing the group numbers of all existing groups
- `Group size`, containing the number of genotypes currently allocated to that group number.

### `see.group.data`
This function returns a vector containing the values of the chosen data type for each group member.

### `see.group.gebvs`
This function returns a dataframe with two columns:

- `i`, containing the indexes of the group members
- `GEBV`, containing the GEBVs of the group member with that index



